lib/utils/threading.jl

```diff
@@ -37,3 +37,52 @@ macro spinlock(l, ex)
     end
   end
 end
+
+
+"""
+    LazyInitialized{T}() do
+        # initialization, producing a value of type T
+    end
+
+A thread-safe, lazily-initialized wrapper for a value of type `T`. Fetch the value by
+calling `getindex`. The constructor is ensured to only be called once from a single thread.
+
+This type is intended for lazy initialization of e.g. global structures, without using
+`__init__`. It is similar to protecting accesses using a lock, but is much cheaper.
+
+"""
+struct LazyInitialized{T,F}
+    # 0: uninitialized
+    # 1: initializing
+    # 2: initialized
+    guard::Threads.Atomic{Int}
```

Comment:
Random thought, we might want to pad this to be on its own cacheline, otherwise two atomic objects could share a cacheline and cause false-sharing.

Reply:
But writes are rare; once the variables have been set-up, the performance hit of false sharing should be negligible, right?

Reply:
Yeah, this was more of a general remark that we might want to allocate the Atomic objects specifically.